std/range/package.d

```diff
@@ -1364,183 +1364,189 @@ Params:
+        @property bool empty()
         {
-            if (condition)
-            {
-                static if (hasElaborateCopyConstructor!R1) r1.__postblit();
-            }
-            else
-            {
-                static if (hasElaborateCopyConstructor!R2) r2.__postblit();
-            }
+            return actOnChosen!(r => r.empty)(this);
         }
 
-        static if (hasElaborateDestructor!R1 || hasElaborateDestructor!R2)
-        ~this()
+    @property auto ref front()
+    {
+        static auto ref getFront(R)(ref R r) { return r.front; }
+        return actOnChosen!getFront(this);
+    }
+
+    void popFront()
+    {
+        return actOnChosen!((ref r) { r.popFront; })(this);
+    }
+
+    static if (isForwardRange!R1 && isForwardRange!R2)
+        @property auto save()
         {
-            if (condition) destroy(r1);
-            else destroy(r2);
+            auto result = this;
+            actOnChosen!((ref r) { r = r.save; })(result);
+            return result;
         }
 
-        static if (isInfinite!R1 && isInfinite!R2)
-            // Propagate infiniteness.
-            enum bool empty = false;
-        else
-            @property bool empty()
-            {
-                return condition ? r1.empty : r2.empty;
-            }
+    @property void front(T)(T v)
+    if (is(typeof({ _r1.front = v; _r2.front = v; })))
+    {
+        actOnChosen!((ref r, T v) { r.front = v; })(this, v);
+    }
 
-        @property auto ref front()
+    static if (hasMobileElements!R1 && hasMobileElements!R2)
+        auto moveFront()
         {
-            return condition ? r1.front : r2.front;
+            return actOnChosen!((ref r) => r.moveFront)(this);
         }
 
-        void popFront()
+    static if (isBidirectionalRange!R1 && isBidirectionalRange!R2)
+    {
+        @property auto ref back()
         {
-            return condition ? r1.popFront : r2.popFront;
+            static auto ref getBack(R)(ref R r) { return r.back; }
+            return actOnChosen!getBack(this);
         }
 
-        static if (isForwardRange!R1 && isForwardRange!R2)
-            @property auto save()
-            {
-                auto result = this;
-                if (condition) r1 = r1.save;
-                else r2 = r2.save;
-                return result;
-            }
-
-        @property void front(T)(T v)
-        if (is(typeof({ r1.front = v; r2.front = v; })))
+        void popBack()
         {
-            if (condition) r1.front = v; else r2.front = v;
+            actOnChosen!((ref r) { r.popBack; })(this);
         }
 
         static if (hasMobileElements!R1 && hasMobileElements!R2)
-            auto moveFront()
+            auto moveBack()
             {
-                return condition ? r1.moveFront : r2.moveFront;
+                return actOnChosen!((ref r) => r.moveBack)(this);
             }
 
-        static if (isBidirectionalRange!R1 && isBidirectionalRange!R2)
+        @property void back(T)(T v)
```

Comment:
same here, if the other back returns ref no need for this

Reply:
ditto